### PR TITLE
[Bug] Fix multimodal pickle storage amplification

### LINKED ISSUE #33388: [Bug] Multimodal tensor views cause 518x pickle memory amplification
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Problem

`get_new_expanded_mm_items()` splits bundled multimodal tensors using views:

```python
def _slice_value(value, start, end):
    if isinstance(value, torch.Tensor):
        return value[start:end]
```

Each image item therefore retains the complete parent storage. When SGLang serializes `mm_inputs` through the default CPU/pickle transport, the parent storage is effectively serialized once per image.

## Reproduction
```python
import pickle
import torch

from sglang.srt.managers.mm_utils import get_new_expanded_mm_items
from sglang.srt.managers.schedule_batch import Modality, MultimodalDataItem

n, rows, width = 512, 8, 64

feature = torch.zeros((n * rows, width), dtype=torch.float32)
grids = torch.tensor([[1, 1, rows]] * n, dtype=torch.int64)

item = MultimodalDataItem(
    modality=Modality.IMAGE,
    feature=feature,
    offsets=[(i, i) for i in range(n)],
    model_specific_data={"image_grid_thw": grids},
)

items = get_new_expanded_mm_items([item])

logical_bytes = sum(
    x.feature.numel() * x.feature.element_size() for x in items
)
pickle_bytes = len(pickle.dumps(items, protocol=pickle.HIGHEST_PROTOCOL))

print("logical MiB:", logical_bytes / 1024**2)
print("pickle MiB:", pickle_bytes / 1024**2)
print("amplification:", pickle_bytes / logical_bytes)
```
### Result

```text
logical MiB: 1.0
pickle MiB: 518.305
amplification: 518.3x
```

This reproduces on the unmodified `origin/kimi-k3` checkout. No custom distributed processor is installed or imported.

## Impact

A production request containing 512 images produced approximately 0.646 GiB of logical features but caused more than 330 GiB of serialization-related memory growth on the tokenizer/rank-0 process, eventually causing OOM.

### Expected behavior

Serialized size should remain proportional to the logical feature size, not:

```text
image_count * parent_storage_size
```

### Suggested fix

Materialize split CPU tensors before pickle transport, including `feature`, `precomputed_embeddings`, and split tensor metadata. The change should avoid unnecessary copies for CUDA IPC or shared-memory transport.

A temporary workaround is:

```python
items = get_new_expanded_mm_items(items)
for item in items:
    if isinstance(item.feature, torch.Tensor):
        item.feature = item.feature.clone()
```

### Environment
```
Python: 3.12.13 (main, Jun 23 2026, 15:18:55) [Clang 22.1.3 ]
CUDA available: True
GPU 0,1,2,3,4,5,6,7: NVIDIA H200
GPU 0,1,2,3,4,5,6,7 Compute Capability: 9.0
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 12.9, V12.9.86
CUDA Driver Version: 570.124.06
PyTorch: 2.11.0+cu129
sglang: 0.0.0.dev1+g15fa94d2c.d20260728
sglang-kernel: 0.4.5+cu129
flashinfer_python: 0.6.15.post1
flashinfer_cubin: 0.6.15.post1
flashinfer_jit_cache: 0.6.15.post1+cu129
triton: 3.6.0
transformers: 5.12.1
torchao: 0.17.0+cu129
numpy: 2.3.5
aiohttp: 3.14.3
fastapi: 0.140.7
huggingface_hub: 1.25.1
interegular: 0.3.3
modelscope: 1.38.1
orjson: 3.11.9
outlines: 0.1.11
packaging: 26.2
psutil: 7.2.2
pydantic: 2.13.4
python-multipart: 0.0.32
pyzmq: 27.1.0
uvicorn: 0.51.0
uvloop: 0.22.1
vllm: Module Not Found
xgrammar: 0.2.1
openai: 2.6.1
tiktoken: 0.13.0
anthropic: 0.120.0
litellm: Module Not Found
torchcodec: 0.11.1+cu129
NVIDIA Topology:
        GPU0  GPU1  GPU2  GPU3  GPU4  GPU5  GPU6  GPU7  CPU Affinity    NUMA Affinity
GPU0     X    NV18  NV18  NV18  NV18  NV18  NV18  NV18  0-23,96-119    0
GPU1    NV18   X    NV18  NV18  NV18  NV18  NV18  NV18  0-23,96-119    0
GPU2    NV18  NV18   X    NV18  NV18  NV18  NV18  NV18  24-47,120-143  1
GPU3    NV18  NV18  NV18   X    NV18  NV18  NV18  NV18  24-47,120-143  1
GPU4    NV18  NV18  NV18  NV18   X    NV18  NV18  NV18  48-71,144-167  2
GPU5    NV18  NV18  NV18  NV18  NV18   X    NV18  NV18  48-71,144-167  2
GPU6    NV18  NV18  NV18  NV18  NV18  NV18   X    NV18  72-95,168-191  3
GPU7    NV18  NV18  NV18  NV18  NV18  NV18  NV18   X    72-95,168-191  3

Legend:
  X    = Self
  NV#  = Connection traversing a bonded set of # NVLinks

NICs:
  mlx5_bond_0, mlx5_bond_1, mlx5_bond_2, mlx5_bond_3, mlx5_bond_4
  mlx5_bond_5, mlx5_bond_6, mlx5_bond_7, mlx5_bond_8, mlx5_bond_9

ulimit soft: 1048576
```

### Related work

- #29465 tracks migrating `mm_inputs` away from `PickleWrapper`.
- #29656 implements msgpack-native `mm_inputs` serialization and may avoid
  this amplification on the msgpack path.

This issue specifically covers the existing pickle path, where tensor views retain and repeatedly serialize their complete parent storage. It also provides a focused regression case that should remain valid while the IPC migration is in progress.

### PR DESCRIPTION
**Pull request by zhujiale520**
## Summary

Fixes #33388.

Multimodal splitting creates per-item tensor views that retain the full parent
storage. Pickling those items serializes the oversized storage once per view,
causing memory usage to scale roughly with `item_count * parent_storage_size`.

This change materializes only CPU tensors whose backing storage is larger than
their logical payload at the pickle boundary. It covers:

- multimodal features
- precomputed embeddings
- nested tensor metadata in `model_specific_data`

CUDA IPC proxies and shared-memory wrappers are not tensors at this boundary,
so those transport paths do not incur additional tensor copies. CPU tensors
that already own right-sized storage are also left unchanged.

## Validation

- Added a CPU regression test covering split features and tensor metadata.
- Added coverage for precomputed embedding views and already-owned tensors.
- Reproduced the pickle ratio locally with the exact helper:
  - before: `65.04x`
  - after: `1.29x`
- `test/registered/unit/managers/test_mm_utils_split.py`: `10 passed`
- `ruff check --select=F401,F821,UP037` passed.
- Python syntax compilation and `git diff --check` passed.

---
### CI States

Latest PR Test (Base): :x: [Run #31253332743](https://github.com/sgl-project/sglang/actions/runs/31253332743)
Latest PR Test (Extra): :x: [Run #31253332640](https://github.com/sgl-project/sglang/actions/runs/31253332640)

### CODE REVIEW

**Comment by smazurov** [comment]
We reproduced this storage amplification on two DGX Spark (GB10) nodes serving GLM-5.3-Flash with TP=2. Before the fix, ten CPU feature views shared 426,935,040 bytes of backing storage in the tokenizer. After pickle IPC, the scheduler held ten separate copies of that storage: 4,269,350,400 bytes. The broadcast payload was about 4.27 GB, and image-bearing requests caused host OOMs.

We have now built and deployed the PR's compaction logic unchanged, retaining its added CPU tests. Our only functional addition to the PR is the send-path integration below. We excluded our earlier compaction extensions and separate receive-buffer copy fix from this image to test this PR's fix on its own.

The PR branch was last updated on August 8, before the native `mm_inputs` migration in #29656 merged on August 20. Our pinned source (`28457f0dcab4ccf748d60f2387a1a4a7fdb5a110`, September 6) includes that migration. Image inputs no longer pass through `wrap_as_pickle`; `TokenizedGenerateReqInput.wrap_pickle_fields` only wraps `time_stats`. The PR therefore needs its integration updated for the newer send path.

The rebased PR passes the CPU test file (11 tests in our pinned tree) even without the integration shim, but then fails a native IPC regression. With the shim, sync/async sends of single and batched requests all pass. The regression uses `get_new_expanded_mm_items`, native request objects and SGLang's actual ZeroMQ helpers, including nested tensor metadata and repeated tensor references:

| Request | Logical feature bytes | Before | After first/second pickle hop |
|---|---:|---:|---:|
| Single | 1,505,280 | 15,063,311 | 1,513,586 |
| Batch of two | 3,010,560 | 30,125,502 | 3,026,052 |

Values and tensor aliases survive both hops. The full GPU gates also passed. The faithful image is serving with the same model settings, including 0.88 static memory fraction and native NEXTN, without runtime profiling. An image-bearing session is running. At 15:02 UTC on September 9, both ranks were running, the head was healthy and generating, swap use was zero, and kernel records since deployment and both serving cgroups showed no OOM events. Our external sampling window so far covers about four minutes of this session, not a long-duration leak test or proof of every final client response.

Could the PR include a native tokenized-request IPC regression alongside its direct `wrap_as_pickle` tests? Those direct tests alone pass even when the newer request path misses compaction.

This is our integration diff on top of the PR rebased onto the September 6 source. It does not change the compaction helper:

```diff
--- a/python/sglang/srt/managers/io_struct.py
+++ b/python/sglang/srt/managers/io_struct.py
@@ -2495,8 +2495,19 @@
     return _maybe_unwrap_pickle(_msgpack_decoder.decode(data))
 
 
+def _materialize_mm_inputs_for_pickle(obj: Any) -> None:
+    # Native mm_inputs no longer pass through wrap_as_pickle after #29656.
+    if isinstance(obj, (BatchTokenizedGenerateReqInput, BatchTokenizedEmbeddingReqInput)):
+        for req in obj.batch:
+            _materialize_mm_inputs_for_pickle(req)
+    elif isinstance(obj, (TokenizedGenerateReqInput, TokenizedEmbeddingReqInput)):
+        if obj.mm_inputs is not None:
+            obj.mm_inputs.materialize_cpu_views_for_pickle()
+
+
 def sock_send(socket: zmq.Socket, obj: Any, flags: int = 0) -> None:
     if _USE_PICKLE_IPC:
+        _materialize_mm_inputs_for_pickle(obj)
         socket.send_pyobj(obj, flags=flags, protocol=pickle.HIGHEST_PROTOCOL)
         return
 
@@ -2513,6 +2524,7 @@
 
 async def async_sock_send(socket: zmq.asyncio.Socket, obj: Any, flags: int = 0) -> None:
     if _USE_PICKLE_IPC:
+        _materialize_mm_inputs_for_pickle(obj)
         await socket.send_pyobj(obj, flags=flags, protocol=pickle.HIGHEST_PROTOCOL)
         return
 
```
